compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -449,6 +453,53 @@ class TreeChecker extends Phase with SymTransformer {
       tree
     }
   }
+
+  /**
+    * Checks that `New` nodes are always wrapped inside `Select` nodes.
+    */
+  private object WrappedNewChecker {
+
+    // Current state of the search. We return as soon as we see the first improperly wrapped `New`.
+    sealed trait State
+    case class NotFound(parent: Option[tpd.Tree]) extends State
+    case class Found(parent: Option[tpd.Tree], child: tpd.New) extends State
```

Comment:
But on the root you use `NotFound`. My comment is only for `Found`.

For `NotFound` you could maybe simplify it using some trick like the `Ticket` or defining a `case object Root extends State`